File: ogn-db/src/documents.rs
use diesel::prelude::*;
use diesel::PgConnection;

use crate::result::Result;
use crate::schema;
use crate::{Document, DocumentId};

pub fn create_document(
    conn: &mut PgConnection,
    id: DocumentId,
    title: &str,
    filetype: &str,
) -> Result<Document> {
    let new_document = Document {
        id,
        title: title.to_string(),
        filetype: filetype.to_string(),
    };
    diesel::insert_into(schema::documents::table)
        .values(&new_document)
        .get_result(conn)
        .map_err(|e| str_err!("Error saving new document {}", e))
}

pub fn get_document(conn: &mut PgConnection, id: DocumentId) -> Result<Option<Document>> {
    schema::documents::table
        .filter(schema::documents::id.eq(id))
        .get_result(conn)
        .optional()
        .map_err(|e| str_err!("Error loading document {}", e))
}

pub fn get_documents(
    conn: &mut PgConnection,
    page_index: i64,
    page_size: i64,
) -> Result<Vec<Document>> {
    schema::documents::table
        .limit(page_size)
        .offset(page_index * page_size)
        .load::<Document>(conn)
        .map_err(|e| str_err!("Error loading documents {}", e))
}

pub fn document_exists(conn: &mut PgConnection, document_id: DocumentId) -> Result<bool> {
    Ok(get_document(conn, document_id)?.is_some())
}

pub fn get_num_documents(conn: &mut PgConnection) -> Result<i64> {
    schema::documents::table
        .select(diesel::dsl::count_star())
        .first(conn)
        .map_err(|e| str_err!("Error loading documents {}", e))
}

pub fn delete_document(conn: &mut PgConnection, document_id: DocumentId) -> Result<usize> {
    diesel::delete(schema::documents::table.filter(schema::documents::id.eq(document_id)))
        .execute(conn)
        .map_err(|e| str_err!("Error deleting document {}", e))
}

File: OgnBackend/src/db/documents.jl
@inline function create_document(pool, document::Document)
    query =
        raw"INSERT INTO documents (id, title, filetype) VALUES ($1, $2, $3) RETURNING id;"
    result = async_execute(pool, query, (document.id, document.title, document.filetype))

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        if isempty(cols.id)
            nothing
        else
            Document(DocumentId(cols.id |> first), document.title, document.filetype)
        end
    end
end

@inline function get_document(pool, document_id::DocumentId)
    query = raw"SELECT * FROM documents WHERE id = $1;"

    result = async_execute(pool, query, (document_id,))

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        if isempty(cols.id)
            nothing
        else
            @assert length(cols.id) == 1
            Document(
                DocumentId(cols.id |> first),
                cols.title |> first,
                cols.filetype |> first,
            )
        end
    end
end

@inline function get_documents(pool)
    query = raw"SELECT * FROM documents;"

    result = async_execute(pool, query)

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        [
            Document(DocumentId(id), title, filetype) for
            (id, title, filetype) in zip(cols.id, cols.title, cols.filetype)
        ]
    end
end

@inline function get_documents(pool, page_num_size)
    page_num, page_size = page_num_size

    query = raw"SELECT * FROM documents LIMIT $1 OFFSET $2;"

    result = async_execute(pool, query, (page_size, page_size * page_num))

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        [
            Document(DocumentId(id), title, filetype) for
            (id, title, filetype) in zip(cols.id, cols.title, cols.filetype)
        ]
    end
end

@inline function get_num_documents(pool)
    query = raw"SELECT COUNT(*) FROM documents;"

    result = async_execute(pool, query)

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        if isempty(cols.count)
            nothing
        else
            @assert length(cols.count) == 1
            cols.count |> first
        end
    end
end

document_exists(pool, document_id::DocumentId) = @async begin
    result = get_document(pool, document_id) |> fetch
    result === nothing ? false : true
end

@inline function delete_document(pool, document_id::DocumentId)
    query = raw"DELETE FROM documents WHERE id = $1;"
    result = async_execute(pool, query, (document_id,))

    @async begin
        result = result |> fetch
        nothing
    end
end

@inline function get_filepath_for_document(document_id::DocumentId)
    path = joinpath(Globals.DB_DOCUMENTS_DIRPATH, "$document_id.pdf")
    isfile(path) ? path : nothing
end
